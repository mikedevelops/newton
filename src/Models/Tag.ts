import { prop, Typegoose } from 'typegoose';

export class Tag extends Typegoose {
    @prop({ required: true })
    name: string;
}

export const TagModel = new Tag().getModelForClass(Tag, {
    schemaOptions: {
        strict: 'throw'
    }
});
