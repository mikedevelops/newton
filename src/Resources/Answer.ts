import mongoose, { Schema } from 'mongoose';

export const Answer = new Schema({
    text: { type: String, required: true }
});

export const AnswerModel = mongoose.model('Answer', Answer);
