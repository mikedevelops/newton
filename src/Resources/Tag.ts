// import { prop, Typegoose } from 'typegoose';
//
// export class Tag extends Typegoose {
//     @prop({ required: true })
//     name: string;
// }
//
// export const TagModel = new Tag().getModelForClass(Tag, {
//     schemaOptions: {
//         strict: 'throw'
//     }
// });

import mongoose, { Schema } from 'mongoose';

export const Tag = new Schema({
    name: { type: String, required: true }
});

export const TagModel = mongoose.model('Tag', Tag);
