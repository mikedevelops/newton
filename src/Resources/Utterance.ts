import mongoose, { Schema } from 'mongoose';

export const Utterance = new Schema({
    text: { type: String, required: true }
});

export const UtteranceModel = mongoose.model('Utterance', Utterance);
