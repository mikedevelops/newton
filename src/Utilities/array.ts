import { randomInRange } from './number';

/**
 * Get a set of random unique indices from an array
 * @param collection
 * @param amount
 */
export const getUniqueRandomIndices = (collection: any[], amount: number) => {
    const indices = [];
    const max = amount >= collection.length ? collection.length : amount;

    for (let index = 0; index < max; index++) {
        const getRandomIndex = randomInRange.bind(this, 0, collection.length - 1);
        let randomIndex;

        do {
            randomIndex = getRandomIndex();
        } while (indices.indexOf(randomIndex) !== -1);

        indices.push(randomIndex);
    }

    return indices;
};
