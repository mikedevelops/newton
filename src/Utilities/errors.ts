import { Model } from 'mongoose';

/**
 * Compose a resource not found error message
 * @param ResourceModel
 * @param resourceId
 */
export const createResourceNotfoundMessage = (ResourceModel: Model<any>, resourceId: string): string => {
    return `Could not find resource "${ResourceModel.modelName}" with ID "${resourceId}"`;
};
