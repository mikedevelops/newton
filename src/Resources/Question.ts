import { Answer } from './Answer';
import { Utterance } from './Utterance';
import { Tag } from './Tag';
import mongoose, { Schema } from 'mongoose';

const Question = new Schema({
    answers: {
        type: [Schema.Types.ObjectId],
        ref: 'Answer'
    },
    utterances: {
        type: [Schema.Types.ObjectId],
        ref: 'Utterance',
        required: true
    },
    tags: {
        type: [Schema.Types.ObjectId],
        ref: 'Tag'
    }
}, {
    strict: 'throw'
});

export const QuestionModel = mongoose.model('Question', Question);
