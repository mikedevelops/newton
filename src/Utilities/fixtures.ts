import { QuestionModel } from '../Resources/Question';
import { UtteranceModel } from '../Resources/Utterance';
import { TagModel } from '../Resources/Tag';
import { AnswerModel } from '../Resources/Answer';
import faker from 'faker';
import { getUniqueRandomIndices } from './array';
import { randomInRange } from './number';
import { Model, Document } from 'mongoose';

/**
 * Compose a createAndSaveEntities entity function
 * @param entityCollectionFactory
 * @param entityPersister
 */
export const createAndSaveEntities = async (
    entityCollectionFactory: () => Promise<any>,
    entityPersister: (entities: any) => Promise<any>
) => {
    return async (...factoryArgs: any[]) => {
        return await entityPersister(await entityCollectionFactory.apply(undefined, factoryArgs));
    };
};

/**
 * Save entities
 * @param entityCollection
 */
export const saveEntities = async (entityCollection: Document[]) => {
    const entities = [];

    for (let index = 0; index < entityCollection.length; index++) {
        const entity = entityCollection[index] as InstanceType<any>;

        entities.push(await entity.save());
    }

    return entities;
};

/**
 * Create entities
 * @param factory
 * @param count
 * @param args
 */
export const createEntities = async (
    factory: () => Document,
    count: number,
    ...args: any[]
) => {
    const entities = [];

    for (let index = 0; index < count; index++) {
        entities.push(await factory.apply(undefined, args));
    }

    return entities;
};

/**
 * Create a Question
 * @param tags
 */
export const createQuestion = async (tags: Document[]) => {
    const questionUtterances = await createEntities(createUtterance, randomInRange(1, 20));
    const questionTags = getUniqueRandomIndices(tags, randomInRange(1, 5)).map(index => tags[index]);
    const answer = await createAnswer().save();

    return new QuestionModel({
        tags: questionTags.map(t => t._id),
        utterances: questionUtterances.map(q => q._id),
        answers: [answer._id]
    });
};

/**
 * Create an Utterance
 */
export const createUtterance = () => {
    return new UtteranceModel({ text: faker.lorem.sentence() });
};

/**
 * Create an Answer
 */
export const createAnswer = () => {
    return new AnswerModel({ text: faker.lorem.paragraph() });
};

/**
 * Create a Tag
 */
export const createTag = () => {
    return new TagModel({ name: faker.commerce.department() });
};

/**
 * Create & Save multiple Question HOF
 * @param count
 */
export const boundCreateAndSaveQuestions = async (count: number) => {
    return await createAndSaveEntities(createEntities.bind(undefined, createQuestion, count), saveEntities);
};

/**
 * Create & Save multiple tags HOF
 * @param count
 */
export const boundCreateAndSaveTags = async (count: number) => {
    return await createAndSaveEntities(createEntities.bind(undefined, createTag, count), saveEntities);
};

