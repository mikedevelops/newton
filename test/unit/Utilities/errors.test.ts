import { createResourceNotfoundMessage } from '../../../src/Utilities/errors';
import { TagModel } from '../../../src/Resources/Tag';

describe('Error Utilities', () => {
    describe('createResourceNotfoundMessage', () => {
        test('should return the error message', () => {
            expect(createResourceNotfoundMessage(TagModel, '666')).toMatchSnapshot();
        });
    });
});
