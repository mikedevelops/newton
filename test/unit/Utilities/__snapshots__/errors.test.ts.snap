// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `"Could not find resource \\"Tag\\" with ID \\"666\\""`;

exports[`Error Utilities createResourceNotfoundMessage should return the error message 1`] = `"Could not find resource \\"Tag\\" with ID \\"666\\""`;
